docs(adr): add decision record for TMX format compliance and validation strategy (#1567)
omegat-org/omegat: src_docs/developer/adr/2025004.TMXFormatAndValidation.md

---

### Option C: Replace with an Alternative Internal Format (e.g., SQLite)
Use a non-TMX format (e.g., SQLite or custom binary XML) for internal storage.
Retain TMX exports for interoperability only.

#### Pros
- Supports rich internal metadata and faster saves
- Clean separation between internal state and exchange formats
- Enables backward-compatible export options
- Addresses performance and scalability concerns raised in discussions

#### Cons
- Loss of human-readable internal format
- Tooling and migration burden
- May require file format version negotiation
- Significant architectural change requiring careful user communication

---

### Option D: Dual Strategy for TMX Validation
Improve TMX handling by adopting **strict schema-based validation** for external files and lenient parsing for trusted internal data.
Use converted **XSD schemas** instead of DTDs for validation.

#### Pros
- Provides reliable TMX file validation across 1.1 and 1.4b versions
- Enhances test coverage and correctness (PR #1559)
- Maintains performance by keeping StAX as the primary parser
- Aligns with XSD-based architecture

#### Cons
- Requires maintenance of XSD schema files

> - Some legacy test files may need to be rewritten

Author: ```suggestion
- Some legacy test files may need to be rewritten (fix is done)
```

2. Validate `project_save.tmx` with TMX 1.1 XSD when reading

#### Pros
- Leverages existing XSD validation infrastructure (PR #1559)
- Provides path for future modernization
- Reduces editor confusion while maintaining compliance

#### Cons
- Questions remain about XML schema references in `<tmx>` element

---

## Implementation Notes

- A pull request [#1559](https://github.com/omegat-org/omegat/pull/1559) has implemented schema-based validation for TMX files using XSDs and demonstrates reading valid TMX without DOCTYPE
- Tests like `AutoTmxTest` and `TMXReaderTest` currently contain files that:
    - Mix versions (e.g., TMX 1.1 with `xml:lang` from 1.4)
    - Omit required attributes (e.g., `pos` on `<it>`)
    - Pass incorrectly due to the absence of validation
- OmegaT's TMXReader uses StAX, which **cannot validate against DTD**; schema validation is the only reliable approach
- Development discussions reveal tension between storage efficiency and exchange format expectations

---

## Recommendation Path

Based on extended discussions, we propose a refined staged approach:

1. **Immediate (Technical Debt Reduction)**

>     - Adopt Option D: Implement XSD-based validation for external TMX files based on PR #1559

Author: ```suggestion
    - Adopt Option D: Implement XSD-based validation for external TMX files proposed as PR #1570
```

Retain TMX exports for interoperability only.

#### Pros
- Supports rich internal metadata and faster saves
- Clean separation between internal state and exchange formats
- Enables backward-compatible export options
- Addresses performance and scalability concerns raised in discussions

#### Cons
- Loss of human-readable internal format
- Tooling and migration burden
- May require file format version negotiation
- Significant architectural change requiring careful user communication

---

### Option D: Dual Strategy for TMX Validation
Improve TMX handling by adopting **strict schema-based validation** for external files and lenient parsing for trusted internal data.
Use converted **XSD schemas** instead of DTDs for validation.

#### Pros
- Provides reliable TMX file validation across 1.1 and 1.4b versions
- Enhances test coverage and correctness (PR #1559)
- Maintains performance by keeping StAX as the primary parser
- Aligns with XSD-based architecture

#### Cons
- Requires maintenance of XSD schema files
- Some legacy test files may need to be rewritten
- Developers must be aware of different validation modes for internal/external files

---

> ### Option E: Hybrid Approach - Remove DOCTYPE, Add XSD Validation
> 1. Remove `DOCTYPE` element while keeping TMX format (`<tmx version="1.1">`)

Author: We can change  "Option E" because OmegaT implement the TMX reader with StAX which does not support DTD validation. So there is no difference whether DOCTYPE exists or not. 
```suggestion
### Option E: Add XSD Validation for both import of TMX data and iinternal data
```

2. Validate `project_save.tmx` with TMX 1.1 XSD when reading

> 3. Research modern XML framework approaches for future writing

Author: ```suggestion
```

In mail list discussion, I agree that the modern XML framework is not fit for TMX data.

#### Pros

> - Eliminates misleading DOCTYPE while maintaining TMX structure

Author: ```suggestion
```

# ADR: Format Compliance and Purpose of `project_save.tmx` and TMX Validation Strategy

## Status
**Proposed** (Date: 2025-07-19)
- Related Threads:
    - "[OmTdev] The internal TMX has DOCTYPE and DTD" (2025-07-13 to 14)
    - "[OmTdev] Some inconsistencies in TMX test cases and format handling" (2025-07-18 to 19)
    - "[OmTdev] Clarifying XML Processing in OmegaT and Use of XSD over DTD" (2025-07-15 to 16)
    - Mailing List: [SourceForge OmegaT-devel July 2025](https://sourceforge.net/p/omegat/mailman/omegat-development/thread/D67B4FB4-137E-4EFB-946C-72BAAB055845%40traductaire-libre.org/)

## Context
OmegaT stores its internal translation memory in a file called `project_save.tmx`. 
Historically, this file has included a DOCTYPE declaration for TMX 1.1 (e.g., `<!DOCTYPE tmx SYSTEM "tmx11.dtd">`).
While this implies TMX conformance, the file is:
- Meant for internal use,

> - Not always valid TMX 1.1 or 1.4b,

Reviewer: Since encapsulation is not a requirement for native codes in TMX, I don’t think we can affirm that the file is not valid TMX 1.1.

Reviewer 2: > Since encapsulation is not a requirement for native codes in TMX, I don’t think we can affirm that the file is not valid TMX 1.1.

Consider it as valid if it makes you happy, but the fact is that any other CAT tool will interpret these tags as part of the text which is not correct. So this file is valid but not correctly readable by other CAT tools.

> - Sometimes misinterpreted by other TMX consumers due to OmegaT-specific tag encoding.

Reviewer 2: In the 1.1 specification:
Each <seg> (Segment) element contains the text of theTUV element.
In the 1.4 specification:
<seg>
Segment - The <seg> element contains the text of the given segment. 

So, if I have a segment like
<seg>&gt;x1/&gt;>sample</seg>
that means that the **text** of the segment is <x1/>sample

---

### Option D: Dual Strategy for TMX Validation
Improve TMX handling by adopting **strict schema-based validation** for external files and lenient parsing for trusted internal data.
Use converted **XSD schemas** instead of DTDs for validation.

#### Pros
- Provides reliable TMX file validation across 1.1 and 1.4b versions
- Enhances test coverage and correctness (PR #1559)
- Maintains performance by keeping StAX as the primary parser
- Aligns with XSD-based architecture

#### Cons
- Requires maintenance of XSD schema files
- Some legacy test files may need to be rewritten
- Developers must be aware of different validation modes for internal/external files

---

### Option E: Hybrid Approach - Remove DOCTYPE, Add XSD Validation
1. Remove `DOCTYPE` element while keeping TMX format (`<tmx version="1.1">`)
2. Validate `project_save.tmx` with TMX 1.1 XSD when reading
3. Research modern XML framework approaches for future writing

#### Pros
- Eliminates misleading DOCTYPE while maintaining TMX structure
- Leverages existing XSD validation infrastructure (PR #1559)
- Provides path for future modernization
- Reduces editor confusion while maintaining compliance

#### Cons

> - Still requires consensus on missing DOCTYPE interpretation

Author: ```suggestion
```
Let's remove a sentence.
